service/storage_service.cc

```diff
@@ -632,6 +632,31 @@ future<> storage_service::topology_change_coordinator_fiber(raft::server& raft,
                 if (!res) {
                     break;
                 }
+
+                // If a node is bootstrapping, we just committed a new CDC generation in the commit_cdc_generation step.
+                // Publish it to the user-facing distributed CDC description tables.
+                if (node.rs->state == node_state::bootstrapping) {
+                    auto curr_gen_id = node.topology->current_cdc_generation_id.value();
+                    auto gen_data = co_await _sys_ks.local().read_cdc_generation(curr_gen_id.id);
+
+                    // Hack: if this is the first node in the cluster, the token ring contains no normal tokens.
```

Comment:
We'd somehow have to add a separate step after the tokens are switched to normal but before we finish the topology change, so basically another state after `write_both_read_new` but before `owner`, where the node is already switched to `node_state::normal` but there's still work to do. The code which picks "node to work on" would have to be modified to also look at normal nodes (?)

I'm not sure it's worth it just to handle this one edge case. If we want to do it I'd leave it for a follow-up.

But generally in the future I'd like to refactor the coordinator a bit and make the topology saga states more explicit, where each step/state of the saga is represented by a separate object. Right now the "next step" is inferred from the states of the nodes. I feel that if we made the "next step" explicit it would be easier to introduce such steps where we need to do some follow-up when the node is already normal etc. And it should also be easier to implement new-cdc-generation-after-node-leaves in this way.

Reply:
> We'd somehow have to add a separate step after the tokens are switched to normal but before we finish the topology change, so basically another state after `write_both_read_new` but before `owner`, where the node is already switched to `node_state::normal` but there's still work to do. The code which picks "node to work on" would have to be modified to also look at normal nodes (?)

I think new state will not be necessary. If we'll move the special case into the `topology_state_load`

> 
> I'm not sure it's worth it just to handle this one edge case. If we want to do it I'd leave it for a follow-up.
> 
> But generally in the future I'd like to refactor the coordinator a bit and make the topology saga states more explicit, where each step/state of the saga is represented by a separate object. Right now the "next step" is inferred from the states of the nodes. I feel that if we made the "next step" explicit it would be easier to introduce such steps where we need to do some follow-up when the node is already normal etc. And it should also be easier to implement new-cdc-generation-after-node-leaves in this way.

The next step has to be inferred from the state and the state alone because a coordinator may change at any moment. Of course the code may create the object from the state and invoke its functions instead of the direct switch, but this is a matter of preference. I prefer less steps since adding a step complicates the recovery.

Reply:
> I think new state will not be necessary. If we'll move the special case into the topology_state_load

I'd prefer to have the coordinator code responsible for publishing the generation, that's where it belongs. On the other hand the coordinator updating token_metadata is also ugly. I don't know which option is uglier to be honest.

> The next step has to be inferred from the state and the state alone because a coordinator may change at any moment.

But not necessarily from *node* states. The next step can be a state by itself.

Reply:
> Both a hacks. We can see how ugly is the one I propose and decide which one is less ugly.

K, I'll try to implement the other hack for v2.

> Yes, but then the same info will be duplicated in to different state: first the replication state and second this new state you propose to have. 

We'd remove replication_state from node state.

Reply:
> > Yes, but then the same info will be duplicated in to different state: first the replication state and second this new state you propose to have.
> 
> We'd remove replication_state from node state.

Well, I need to see it to understand what you mean. May be you just want to organize the state differently.

Reply:
> K, I'll try to implement the other hack for v2.

@gleb-cloudius I tried this:
```diff
diff --git a/cdc/generation_service.hh b/cdc/generation_service.hh
index 5d2a342945..b213080312 100644
--- a/cdc/generation_service.hh
+++ b/cdc/generation_service.hh
@@ -144,6 +144,13 @@ class generation_service : public peering_sharded_service<generation_service>
      */
     future<> handle_cdc_generation(cdc::generation_id_v2);
 
+    /* generation_service code might be racing with system_distributed_keyspace deinitialization
+     * (the deinitialization order is broken).
+     * Therefore, whenever we want to access sys_dist_ks in a background task,
+     * we need to check if the instance is still there. Storing the shared pointer will keep it alive.
+     */
+    shared_ptr<db::system_distributed_keyspace> get_sys_dist_ks();
+
 private:
     /* Retrieve the CDC generation which starts at the given timestamp (from a distributed table created for this purpose)
      * and start using it for CDC log writes if it's not obsolete.
@@ -181,13 +188,6 @@ class generation_service : public peering_sharded_service<generation_service>
      */
     future<> legacy_scan_cdc_generations();
 
-    /* generation_service code might be racing with system_distributed_keyspace deinitialization
-     * (the deinitialization order is broken).
-     * Therefore, whenever we want to access sys_dist_ks in a background task,
-     * we need to check if the instance is still there. Storing the shared pointer will keep it alive.
-     */
-    shared_ptr<db::system_distributed_keyspace> get_sys_dist_ks();
-
     /* Part of the upgrade procedure. Useful in case where the version of Scylla that we're upgrading from
      * used the "cdc_streams_descriptions" table. This procedure ensures that the new "cdc_streams_descriptions_v2"
      * table contains streams of all generations that were present in the old table and may still contain data
diff --git a/service/storage_service.cc b/service/storage_service.cc
index f67d137cca..34fbd93030 100644
--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -407,6 +407,19 @@ future<> storage_service::topology_state_load(cdc::generation_service& cdc_gen_s
     if (auto gen_id = _topology_state_machine._topology.current_cdc_generation_id) {
         slogger.debug("topology_state_load: current CDC generation ID: {}", gen_id);
         co_await cdc_gen_svc.handle_cdc_generation(*gen_id);
+
+        // Hack: if this is the first node in the cluster, the topology coordinator wasn't able
+        // to publish the generation to the stream description tables (token_metadata wasn't ready
+        // yet). Do it here instead.
+        // FIXME: this will be done each time the node restarts, how to fix it?
+        // Also how to handle write failure?
+        if (_topology_state_machine._topology.normal_nodes.size() == 1) {
+            auto gen_data = co_await _sys_ks.local().read_cdc_generation(gen_id->id);
+            auto sys_dist_ks = cdc_gen_svc.get_sys_dist_ks();
+            co_await sys_dist_ks->create_cdc_desc(
+                gen_id->ts, gen_data, { get_token_metadata().count_normal_token_owners() });
+            slogger.info("topology_state_load: published generation {} to description tables", *gen_id);
+        }
     }
 }
 
@@ -645,22 +658,17 @@ future<> storage_service::topology_change_coordinator_fiber(raft::server& raft,
                     auto curr_gen_id = node.topology->current_cdc_generation_id.value();
                     auto gen_data = co_await _sys_ks.local().read_cdc_generation(curr_gen_id.id);
 
-                    // Hack: if this is the first node in the cluster, the token ring contains no normal tokens.
+                    // Hack: if this is the first node in the cluster, the token ring contains no normal tokens
+                    // yet, so we cannot do the 'distributed' write to CDC description tables.
+                    // The write will be done by `topology_state_load` in that case.
                     // Modify token_metadata, setting our tokens to normal so we can do the 'distributed' write for CDC.
                     // It doesn't break anything to set tokens to normal early in this single-node case.
                     bool is_first_node = node.topology->normal_nodes.empty();
-                    assert(!is_first_node || (raft.id() == node.id && node.topology->transition_nodes.size() == 1));
-                    if (is_first_node) {
-                        auto& tokens = node.rs->ring.value().tokens;
-                        co_await mutate_token_metadata([this, &tokens] (mutable_token_metadata_ptr tmptr) {
-                            auto my_address = get_broadcast_address();
-                            tmptr->update_topology(my_address, _sys_ks.local().local_dc_rack());
-                            return tmptr->update_normal_tokens(tokens, my_address);
-                        });
+                    if (!is_first_node) {
+                        co_await sys_dist_ks.local().create_cdc_desc(
+                            curr_gen_id.ts, gen_data, { get_token_metadata().count_normal_token_owners() });
+                        slogger.info("raft topology: published generation {} to description tables", curr_gen_id);
                     }
-
-                    co_await sys_dist_ks.local().create_cdc_desc(
-                        curr_gen_id.ts, gen_data, { get_token_metadata().count_normal_token_owners() });
                 }
 
                 raft_topology_cmd cmd{raft_topology_cmd::command::stream_ranges};
```

Unfortunately, when I tried to bootstrap the third node, it failed with
```
ERROR 2023-04-05 14:59:12,226 [shard 0] init - Startup failed: raft::stopped_error (Raft instance is stopped, reason: "background error, std::_Nested_exception<raft::state_machine_error> (State machine error at raft/raft.hh:299): exceptions::unavailable_exception (Cannot achieve consistency level for cl ONE. Requires 1, alive 0)")
```

IIUC it tried to perform the write when it was replaying the Raft log, reloading the topology after each command. After applying the command which turned the original (first) node to normal, the booting node tried to perform the write and failed (probably because its topology was "wrong" at this point?)

Managed to fix it with more complex condition:
```diff
+        auto& normal_nodes = _topology_state_machine._topology.normal_nodes;
+        if (normal_nodes.size() == 1 && normal_nodes.begin()->first == _group0->group0_server().id()) {
```
maybe there's a simpler condition we could use...

Anyway I very much dislike this hack:
- the first node will do this write every time the topology transitions as long as it's the only node in the cluster (on restarts for example). This is wasteful. After booting 2-node cluster I saw this write happen 5 times in the logs. Maybe there's a smarter condition though.
- we had to introduce a dependency on `system_distributed_keyspace` in `topology_state_load`. `topology_state_load` shouldn't have to contact other nodes, only read from local tables and update local data structures IMO.
- it's also unclear how to handle a write failure here. The coordinator simply retries on errors, but what should we do if `topology_state_load` fails?
- the logic of the hack is scattered across the code, we still need the `if` inside coordinator code that checks if we're not the first node and we have do the write in two places now.

Reply:
Done